### PR TITLE
Implement fee_discount field, fix several bugs in claim_account_evaluator #2651

### PR DESCRIPTION
**Pull request by theoreticalbts**
PR for #2651

### CODE REVIEW

File: libraries/chain/steem_evaluator.cpp [hunk 1]
```diff
@@ -161,6 +166,11 @@ void witness_set_properties_evaluator::do_apply( const witness_set_properties_op
    if( itr != o.props.end() )
    {
       fc::raw::unpack_from_vector( itr->second, props.account_creation_fee );
+      if( _db.has_hardfork( STEEM_HARDFORK_0_20__2651 ) || _db.is_producing() )
```
Comment (mvandeberg) [inline libraries/chain/steem_evaluator.cpp]: This is a new op. The validation can be moved the the validate method now.

**Comment by theoreticalbts** [comment]
Rebased against `develop` to resolve conflicts

**Comment by theoreticalbts** [comment]
Pushed, hopefully this will fix the CI build

File: libraries/chain/steem_evaluator.cpp [hunk 2]
```diff
@@ -2335,19 +2357,61 @@ void claim_account_evaluator::do_apply( const claim_account_operation& o )
...
 
-   const auto& creation_fee = wso.median_props.account_creation_fee;
-
-   asset paid_fee = o.fee;
-
-   if( paid_fee < creation_fee )
-   {
-      // This is optimized because STEEM_100_PERCENT == STEEM_ACCOUNT_SUBSIDY_PRECISION
-      uint64_t percent_unpaid = ( ( creation_fee.amount.value - paid_fee.amount.value ) * STEEM_100_PERCENT ) / creation_fee.amount.value;
+   FC_TODO( "Remove min() after HF20" );
```
Comment (mvandeberg) [inline libraries/chain/steem_evaluator.cpp]: This op is not active until HF 20 so the check is not needed.

File: libraries/chain/steem_evaluator.cpp [hunk 3]
```diff
@@ -2358,20 +2422,11 @@ void claim_account_evaluator::do_apply( const claim_account_operation& o )
          uint32_t delta_time = ( _db.head_block_time() - current_witness.last_subsidy_update ).to_seconds();
 
          // linear decay to enforce no more than 10% max by a single witness over 48 hours.
-         uint64_t recovered_subsidies = ( delta_time * STEEM_100_PERCENT * wso.single_witness_subsidy_limit )
-            / ( fc::days( STEEM_ACCOUNT_SUBSIDY_BURST_DAYS ).to_seconds() * STEEM_100_PERCENT );
-         uint64_t new_subsidies = current_witness.recent_account_subsidies;
+         uint64_t recovered_subsidies = ( uint64_t( delta_time ) * uint64_t( wso.single_witness_subsidy_limit ) )
+                / (STEEM_ACCOUNT_SUBSIDY_BURST_DAYS * 60 * 60 * 24);
```
Comment (mvandeberg) [inline libraries/chain/steem_evaluator.cpp]: Despite making the pool size votable is this still a reasonable heuristic to limit witness usage?

File: libraries/chain/steem_evaluator.cpp [hunk 4]
```diff
@@ -2335,19 +2357,61 @@ void claim_account_evaluator::do_apply( const claim_account_operation& o )
...
-      // This is optimized because STEEM_100_PERCENT == STEEM_ACCOUNT_SUBSIDY_PRECISION
-      uint64_t percent_unpaid = ( ( creation_fee.amount.value - paid_fee.amount.value ) * STEEM_100_PERCENT ) / creation_fee.amount.value;
+   FC_TODO( "Remove min() after HF20" );
+   int64_t creation_fee = std::min( wso.median_props.account_creation_fee.amount.value, STEEM_MAX_ACCOUNT_CREATION_FEE );
+
+   FC_ASSERT( o.fee.amount.value <= creation_fee, "Cannot pay more than account creation fee. paid: ${p} fee: ${f}", ("p", o.fee.amount.value)("f", creation_fee) );
+
+   int64_t fee_discount = creation_fee - o.fee.amount.value;
+
+   if( fee_discount > 0 )
```
Comment (mvandeberg) [inline libraries/chain/steem_evaluator.cpp]: I thought the intent behind this change was to make it so the the fee was wither the creation fee or it was 0.

**Comment by theoreticalbts** [comment]
I'm going to add another commit here to remove partial subsidization.

**Comment by theoreticalbts** [comment]
In the interests of moving this forward, I'm going to:

- Move the refactoring commits to a new branch
- Rewind this PR (using `git push -f`) to only contain the code to remove the `fee_discount` field, in line with the title of the PR 
- Fix any merge conflicts

**Comment by theoreticalbts** [comment]
Finished, ready for review.

**Comment by mvandeberg** [comment]
This is bug fix only.

**Pull request merged**